File: Python/6py 2/roman_decoder.py
"""
|--------------------------------------------------------------------------
| Roman Numerals Decoder
|--------------------------------------------------------------------------
|
"""

def solution(roman):
   x = {'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}
   res, prev= 0, 0
   for char in roman:
      res += x[char]
      if x[char] > prev: 
         res -= 2 * prev
      prev = x[char]
   return res


def solution(roman):
   values = [('M', 1000), ('CM', -200), ('D', 500), ('CD', -200),
         ('C', 100), ('XC', -20), ('L', 50), ('XL', -20),
         ('X', 10), ('IX', -2), ('V', 5), ('IV', -2),
         ('I', 1)]
   return sum(roman.count(s)*v for s,v in values)


def solution(roman): 
   inter = [ {'I':1,'V':5,'X':10,'L':50,'C':100,'D':500,'M':1000 }[i] for i in roman ]+[0]       
   return sum( [-inter[i] if inter[i]<inter[i+1] else inter[i] for i in range(len(inter)-1) ])


print(solution("XXI"), 21)
print(solution("I"), 1)
print(solution("IV"), 4)
print(solution("MMVIII"), 2008)
print(solution("MDCLXVI"), 1666)

File: JavaScript/6js 3/roman_decoder.js
/*
|--------------------------------------------------------------------------
| Roman Numerals Decoder
|--------------------------------------------------------------------------
|
*/

function solution(roman) {
   const x = {
      I: 1,
      V: 5,
      X: 10,
      L: 50,
      C: 100,
      D: 500,
      M: 1000,
   };
   let res = 0;
   for (let i = 0; i < roman.length; i++) {
      const current = x[roman[i]];
      const next = x[roman[i + 1]];
      if (next && current < next) {
         res += next - current;
         i++;
      } else {
         res += current;
      }
   }
   return res;
}


function solution(roman) {
   let conversion = {
      M: 1000,
      CM: 900,
      D: 500,
      CD: 400,
      C: 100,
      XC: 90,
      L: 50,
      XL: 40,
      X: 10,
      IX: 9,
      V: 5,
      IV: 4,
      I: 1,
   };
   return roman
      .match(/CM|CD|XC|XL|IX|IV|\w/g)
      .reduce((accum, roman) => accum + conversion[roman], 0);
}


function solution(roman) {
   let value = { I: 1, V: 5, X: 10, L: 50, C: 100, D: 500, M: 1000 };
   return roman
      .split("")
      .map((d) => value[d])
      .reduce((s, v, i, o) => s + (o[i + 1] > v ? -v : v), 0);
}


console.log(solution("XXI"), 21);
console.log(solution("I"), 1);
console.log(solution("IV"), 4);
console.log(solution("MMVIII"), 2008);
console.log(solution("MDCLXVI"), 1666);